### PR TITLE
Lightning UI

### PR DESCRIPTION
**Pull request by benjaminpkane**
Introduces new `FIFTYONE_APP_LIGHTNING_THRESHOLD` option that is `None` by default. It accepts a positive integer value. When a dataset's sample count exceeds that value and there is no view present, lightning mode is enabled in the App.

In lightning mode, only indexed fields are filterable until the filtered count falls below the lightning threshold

Indexed filters filter have the following constraints
* counts do not display for string and boolean fields
* numeric fields are not displayed when nonfinite values are present
* id fields do not provide search results

These constraints enable fast queries (IXSCAN and DISTINCT) via #3667 

After filtering yields a sample count below the threshold, all other filters are accessible either through `see more` (for label fields with indexed subfields, or the top-level caret. Counts are also presented 

### Changes

I've added some hopefully improved organization to the `@fiftyone/core` and `@fiftyone/state` packages to introduce new data flow when `lightning` mode is enabled.

Notable code locations

### @fiftyone/core

* [LightningFilterablePathEntries.tsx](https://github.com/voxel51/fiftyone/blob/lightning-ui/app/packages/core/src/components/Sidebar/Entries/FilterablePathEntry/LightningFilterablePathEntries.tsx)
* [Filters](https://github.com/voxel51/fiftyone/tree/lightning-ui/app/packages/core/src/components/Filters)

### @fiftyone/state

* [pathData](https://github.com/voxel51/fiftyone/tree/lightning-ui/app/packages/state/src/recoil/pathData)
* [lightning.ts](https://github.com/voxel51/fiftyone/blob/lightning-ui/app/packages/state/src/recoil/lightning.ts)
* [sidebarExpanded.ts](https://github.com/voxel51/fiftyone/blob/lightning-ui/app/packages/state/src/recoil/sidebarExpanded.ts)

Considerations I propose we include in separate PRs

* App action row setting for disabling lightning mode and/or adjusting the threshold
* UX for `see more` filters 
* Combine legacy sidebar mode (fast, best, all) and lightning

### TODO

* [x] passing e2e
* [x] lightning description in field info tooltip
* [x] errors states for too many results in StringFilter

### Testing

Configure
```py
export FIFTYONE_APP_LIGHTNING_THRESHOLD=199
```

Global wildcard index
```py
import fiftyone.zoo as foz

image = foz.load_zoo_dataset("quickstart")

# wildcard indexes only enable non-numeric filters
image.create_index([("$**", 1)])
```

Edges cases
```py
import fiftyone as fo

fo.delete_dataset("lightning-demo")
dataset = fo.Dataset("lightning-demo")

dataset.add_sample(
    fo.Sample(
        filepath=f"nonfinites.png",
        inf=float("inf"),
        nan=float("nan"),
        ninf=float("-inf"),
    )
)

for i in range(0, 100):
    dataset.add_sample(
        fo.Sample(filepath=f"{i}.png", inf=i, nan=i, ninf=i, str=str(i))
    )

dataset.add_sample(fo.Sample(filepath="none.png", str=None))
for i in ["inf", "nan", "ninf", "str"]:
    dataset.create_index(i)
```

https://github.com/voxel51/fiftyone/assets/19821840/f12b33b7-1966-4dce-b0dc-f47d8f54e41d

### CODE REVIEW

File: app/packages/core/src/components/Sidebar/Entries/FilterablePathEntry/FilterablePathEntry.tsx [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+          {!disabled && !(modal && path === LABEL_TAGS) && (
+            <Checkbox
+              checked={active}
+              title={`Show ${path}`}
+              style={{
+                color: active ? color : theme.text.secondary,
+                marginLeft: 2,
+                padding: 0,
+              }}
+              key="checkbox"
```
Comment (lanzhenw) [inline app/packages/core/src/components/Sidebar/Entries/FilterablePathEntry/FilterablePathEntry.tsx:89]: can key be more specific?

File: app/packages/core/src/components/Filters/FilterOption/useOptions.ts [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      ? path.split(".").slice(-2)[0]
+      : null;
+
+    const options: Option[] = [];
+    if (!isFilterMode) {
+      options.push({
+        icon: "VisibilityIcon",
+        key: "visible",
+        value: `Show ${name} 
+        `,
```
Comment (lanzhenw) [inline app/packages/core/src/components/Filters/FilterOption/useOptions.ts:51]: extra line

File: app/packages/core/src/components/Filters/FilterOption/FilterItem.tsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            style={{ display: "flex", flexDirection: "row" }}
+            ref={ref}
+            onClick={onClick}
+            data-cy={`filter-option-${id}`}
+          >
+            {icon ? (
+              <IconButton sx={{ color: color }}>{getIcon(icon!)}</IconButton>
+            ) : null}
+
+            <Text>{value}</Text>
```
Comment (lanzhenw) [inline app/packages/core/src/components/Filters/FilterOption/FilterItem.tsx:96]: Screenshot 2023-11-16 at 12 41 24 AM
visibility mode did not display selected option

**Review by lanzhenw — CHANGES_REQUESTED** [review]
🚀 Great work to split logic and components out of the sidebar filter. ty for adding a lot more selectors. Lightning mode FE flow is well organized.

Left some comments for minor issues.

File: app/packages/components/src/components/Results/Results.tsx [hunk 4]
```diff
@@ -99,18 +98,16 @@ const Results = <T extends unknown>({
...
-          {!total && <>No results</>}
-        </div>
-      )}
+      <div className={style.footer}>
+        {total && results.length && (
+          <>
+            {results.length} of {total.toLocaleString()}
+          </>
+        )}
+        {!results.length && <>No results</>}
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/components/src/components/Results/Results.tsx:107]: I think `results` should always be empty when `total` is 0

File: app/packages/core/src/components/Filters/BooleanFieldFilter.tsx [hunk 5]
```diff
@@ -5,35 +5,32 @@ import {
...
         path,
         modal,
         extended: false,
       })}
       modal={modal}
       path={path}
-      color={color}
       {...rest}
     />
   );
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/core/src/components/Filters/BooleanFieldFilter.tsx:36]: Fixed!

I want to add some e2e tests after the dev cut

File: app/packages/core/src/components/Sidebar/Entries/FilterablePathEntry/FilterablePathEntry.tsx (hunk 1 above)
Comment (benjaminpkane) [reply app/packages/core/src/components/Sidebar/Entries/FilterablePathEntry/FilterablePathEntry.tsx]: This a just so React doesn't complain. It's not new code 😅 . But I also added a key for its sibling element

**Review by manivoxel51 — APPROVED** [review]
Great work! 🍨 nice refactor!

besides couple of nits, here is my suggestion

### Major issue to investigate

The embeddings don't seem to work on my end. selecting a brain key has no effect.

https://github.com/voxel51/fiftyone/assets/109545780/81a96bcb-d27f-49de-858a-8845ebc4b20f


UX feedback - can come later

- There will be confusion for users who set the threshold, then try to filter using the sidebar and seeing "disabled" with colored downward carot without an indication as to why is it disabled and how can I reset the state so I can filter. One suggestion is to 
1. gray out the carot on the field that is disabled (keeping it colored is fine)
2. show a button somewhere in the header so they can reset the threshold to none
3. There are certain element(s) like Tagger that disappear in the lighning mode. could be confusing and raise lot of "Where is the Tagger?". My suggestion is to disable it and add a tooltip "Disabled - in lighning mode" or similar.

Question
-

File: app/packages/core/src/components/Filters/FilterOption/useOnSelect.ts [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    excluded && setExcluded(false);
+    !isMatching && setIsMatching(true);
+  };
+
+  const onSelectNegativeMatch = () => {
+    !excluded && setExcluded(true);
+    !isMatching && setIsMatching(true);
+  };
+
+  const onSelectVisible = () => {
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/core/src/components/Filters/FilterOption/useOnSelect.ts:62]: Not new code. Trying to avoid bugs

**Comment by benjaminpkane** [comment]
> gray out the carot on the field that is disabled (keeping it colored is fine)

This is already the case. We can refine post dev cut 👍

File: app/packages/components/src/components/Results/Results.tsx [hunk 7]
```diff
@@ -85,32 +86,32 @@ const Results = <T extends unknown>({
...
+              active={i === active}
+              component={component}
+              key={toKey(result)}
+              result={result}
+              onClick={() => onSelect(result)}
+            />
+          ))}
+      </div>
+      <div className={style.footer}>
+        {!results && !!noResults && <>{noResults}</>}
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/components/src/components/Results/Results.tsx:105]: I will refine post dev cut. The missing behavior `noResults` is trying to resolve is an error message state

File: app/packages/state/src/recoil/utils.ts [hunk 8]
```diff
@@ -163,6 +163,23 @@ export const getEmbeddedLabelFields = (fields: StrictField[], prefix = "") =>
     )
     .flat();
 
+export type Range = [number | null | undefined, number | null | undefined];
+
+export const asDefaultRange = (
```
[the remark this answers is not in the document]
Comment (benjaminpkane) [reply app/packages/state/src/recoil/utils.ts:168]: I'd like to remove the default range altogether. It causes so many bugs

**Review by lanzhenw — APPROVED** [review]

**Pull request merged**